### PR TITLE
Fix typing in SdoTermSource

### PR DESCRIPTION
**Pull request by MatthiasWiesmann**
Added typing and structure to SdoTermSource.

The various fields of SdoTerm contain either identifiers (as strings), or instances of SdoTerm and the code hacks around this with a lot of obscure casting operations. Replaced the content with an object than contains either, or a sequence which contains either. This makes it possible to at least **document** the types that are taken by various methods. This also enables, making the code smarter, as most of the current functions systematically assume the worst. 

* Added holding structures.
* Added unit-tests for these structures.
* Updated all the code that access theses fields, including the Jinga templates.
* Added typing information in many places.
* Removed test that was broken but shadowed by another test with the same name (thanks python). 
* Added documentation where I figured out what is going on.
* Added some assertions on types in critical entry points.

### CODE REVIEW

File: software/SchemaTerms/sdoterm.py [hunk 1]
```diff
@@ -17,65 +20,257 @@ class SdoTermType(str, enum.Enum):
...
+                term_id = element.id
+            except AttributeError:
+                term_id = element
+            term_ids.append(term_id)
+        sequence.setIds(term_ids)
+        return sequence
+
+    @property
+    def expanded(self):
+        return not self._term_dict or all(self._term_dict.values())
```
Comment (pascalfleury) [inline software/SchemaTerms/sdoterm.py:102]: `all` returns `True` when the iterable is empty, and the iterable is set in `__init__` so the first part of the test is not needed?

File: software/SchemaTerms/sdoterm.py [hunk 2]
```diff
@@ -17,65 +20,257 @@ class SdoTermType(str, enum.Enum):
...
+
+
+class SdoTerm(object):
+    """Abstract superclass for various schema.org types.
+
+    Note that the semantics of the relational fields depends on the expansion_depth.
+
+    0: Nothing is expanded
+    1: Non recursive fields are expanded
+    2: All recursive fields are expanded (at are at least at level 1)
```
Comment (pascalfleury) [inline software/SchemaTerms/sdoterm.py:147]: at-> and ?

**Review by pascalfleury — APPROVED** [review]

File: software/SchemaTerms/sdoterm.py [hunk 3]
```diff
@@ -17,65 +20,257 @@ class SdoTermType(str, enum.Enum):
...
 
+class UnexpandedTermError(LookupError):
+    """Term is not expanded."""
+
+
+class SdoTermOrId(object):
+    """Wrapper that holds a term-id or a term."""
+
+    def __init__(self, term_id : str = None, term = None):
+        self._term = term
```
[the remark this answers is not in the document]
Comment (MatthiasWiesmann) [reply software/SchemaTerms/sdoterm.py:31]: Made it a xor condition.

File: software/SchemaTerms/sdoterm.py [hunk 4]
```diff
@@ -17,65 +20,257 @@ class SdoTermType(str, enum.Enum):
...
+        self._term_id = term_id
+
+    def setTerm(self, term):
+        self._term = term
+        if self._term:
+            self._term_id = term.id
+
+    def __str__(self):
+        if not self._term:
+            return '<%s>' % self._term_id
```
[the remark this answers is not in the document]
Comment (MatthiasWiesmann) [reply software/SchemaTerms/sdoterm.py:62]: Done.

File: software/SchemaTerms/sdoterm.py [hunk 5]
```diff
@@ -17,65 +20,257 @@ class SdoTermType(str, enum.Enum):
...
+        return str(self._term)
+
+    def __bool__(self):
+        return bool(self._term_id)
+
+
+class SdoTermSequence(object):
+    """Sequence that holds either a sequence of term-ids, or a sequence of terms.
+
+    If it holds a sequence of terms, the sequence is said to expanded.
```
[the remark this answers is not in the document]
Comment (MatthiasWiesmann) [reply software/SchemaTerms/sdoterm.py:72]: Done.

File: software/SchemaTerms/sdoterm.py (hunk 1 above)
Comment (MatthiasWiesmann) [reply software/SchemaTerms/sdoterm.py:102]: Done.

File: software/SchemaTerms/sdoterm.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (MatthiasWiesmann) [reply software/SchemaTerms/sdoterm.py:31]: Actually, the object can hold None, so I changed the initalisation to accept either one, or none. 
Updated the documentation to reflect this.

**Pull request merged**